common/src/main/scala/org/apache/celeborn/common/meta/PartitionLocationInfo.scala

```diff
@@ -75,24 +93,39 @@ class PartitionLocationInfo extends Logging {
   }
 
   def getAllMasterLocations(shuffleKey: String): util.List[PartitionLocation] = this.synchronized {
-    if (masterPartitionLocations.containsKey(shuffleKey)) {
-      masterPartitionLocations.get(shuffleKey)
-        .values()
-        .asScala
-        .flatMap(_.asScala)
-        .toList
-        .asJava
-    } else {
-      new util.ArrayList[PartitionLocation]()
-    }
+    getMasterLocations(shuffleKey)
   }
 
   def getAllSlaveLocations(shuffleKey: String): util.List[PartitionLocation] = this.synchronized {
-    if (slavePartitionLocations.containsKey(shuffleKey)) {
-      slavePartitionLocations.get(shuffleKey)
+    getSlaveLocations(shuffleKey)
+  }
+
+  def getMasterLocations(
+      shuffleKey: String,
+      partitionIdOpt: Option[Int] = None): util.List[PartitionLocation] = {
+    getLocations(shuffleKey, masterPartitionLocations, partitionIdOpt)
+  }
+
+  def getSlaveLocations(
+      shuffleKey: String,
+      partitionIdOpt: Option[Int] = None): util.List[PartitionLocation] = {
+    getLocations(shuffleKey, slavePartitionLocations, partitionIdOpt)
+  }
+
+  private def getLocations(
+      shuffleKey: String,
+      partitionInfo: PartitionInfo,
+      partitionIdOpt: Option[Int] = None): util.List[PartitionLocation] = this.synchronized {
+    if (partitionInfo.containsKey(shuffleKey)) {
+      partitionInfo.get(shuffleKey)
         .values()
         .asScala
         .flatMap(_.asScala)
+        .filter(p =>
```

Comment:
It's inefficient to check for each element. Better to put the if in the beginning.